src/VisualStudio/Core/Def/Implementation/ProjectSystem/AbstractProject.cs

```diff
@@ -1096,12 +1098,48 @@ internal void RemoveAdditionalDocument(IVisualStudioHostDocument document)
             lock (_gate)
             {
                 _additionalDocuments.Remove(document.Id);
-                _documentMonikers.Remove(document.Key.Moniker);
+                RemoveMoniker(document);
             }
 
             UninitializeAdditionalDocument(document);
         }
 
+        private void AddMoniker(IVisualStudioHostDocument document)
+        {
+            var moniker = document.Key.Moniker;
+            if (_documentMonikers.TryGetValue(moniker, out var value))
+            {
+                value.refCount++;
+                _documentMonikers[moniker] = (value.document, value.refCount);
+            }
+            else
+            {
+                _documentMonikers.Add(moniker, (document, 1));
+            }
+        }
+
+        private void RemoveMoniker(IVisualStudioHostDocument document)
+        {
+            var moniker = document.Key.Moniker;
+            if (_documentMonikers.TryGetValue(moniker, out var value))
+            {
+                Debug.Assert(value.document.Equals(document));
+                value.refCount--;
+                if (value.refCount == 0)
+                {
+                    _documentMonikers.Remove(moniker);
+                }
+                else
+                {
+                    _documentMonikers[moniker] = (value.document, value.refCount);
```

Comment:
Can't the right hand side just be "value" since you already have the tuple in hand though?